{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0ed52b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "import graphviz\n",
    "import itertools\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b918028",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DeBruijnGraph:\n",
    "    ''' De Bruijn directed multigraph built from a collection of\n",
    "        strings. User supplies strings and k-mer length k.  Nodes\n",
    "        are k-1-mers.  An Edge corresponds to the k-mer that joins\n",
    "        a left k-1-mer to a right k-1-mer. '''\n",
    " \n",
    "    @staticmethod\n",
    "    def chop(st, k):\n",
    "        ''' Chop string into k-mers of given length '''\n",
    "        for i in range(len(st)-(k-1)):\n",
    "            yield (st[i:i+k], st[i:i+k-1], st[i+1:i+k])\n",
    "    \n",
    "    class Node:\n",
    "        ''' Node representing a k-1 mer.  Keep track of # of\n",
    "            incoming/outgoing edges so it's easy to check for\n",
    "            balanced, semi-balanced. '''\n",
    "        \n",
    "        def __init__(self, km1mer):\n",
    "            self.km1mer = km1mer\n",
    "            self.nin = 0\n",
    "            self.nout = 0\n",
    "        \n",
    "        def isSemiBalanced(self):\n",
    "            return abs(self.nin - self.nout) == 1\n",
    "        \n",
    "        def isBalanced(self):\n",
    "            return self.nin == self.nout\n",
    "        \n",
    "        def __hash__(self):\n",
    "            return hash(self.km1mer)\n",
    "        \n",
    "        def __str__(self):\n",
    "            return self.km1mer\n",
    "    \n",
    "    def __init__(self, strIter, k, circularize=False):\n",
    "        ''' Build de Bruijn multigraph given string iterator and k-mer\n",
    "            length k '''\n",
    "        self.G = {}     # multimap from nodes to neighbors\n",
    "        self.nodes = {} # maps k-1-mers to Node objects\n",
    "        for st in strIter:\n",
    "            if circularize:\n",
    "                st += st[:k-1]\n",
    "            for kmer, km1L, km1R in self.chop(st, k):\n",
    "                nodeL, nodeR = None, None\n",
    "                if km1L in self.nodes:\n",
    "                    nodeL = self.nodes[km1L]\n",
    "                else:\n",
    "                    nodeL = self.nodes[km1L] = self.Node(km1L)\n",
    "                if km1R in self.nodes:\n",
    "                    nodeR = self.nodes[km1R]\n",
    "                else:\n",
    "                    nodeR = self.nodes[km1R] = self.Node(km1R)\n",
    "            \n",
    "                nodeL.nout += 1\n",
    "                nodeR.nin += 1\n",
    "                self.G.setdefault(nodeL, []).append(nodeR)\n",
    "        # Iterate over nodes; tally # balanced, semi-balanced, neither\n",
    "        self.nsemi, self.nbal, self.nneither = 0, 0, 0\n",
    "        # Keep track of head and tail nodes in the case of a graph with\n",
    "        # Eularian walk (not cycle)\n",
    "        self.head, self.tail = None, None\n",
    "        for node in iter(self.nodes.values()):\n",
    "            if node.isBalanced():\n",
    "                self.nbal += 1\n",
    "            elif node.isSemiBalanced():\n",
    "                if node.nin == node.nout + 1:\n",
    "                    self.tail = node\n",
    "                if node.nin == node.nout - 1:\n",
    "                    self.head = node\n",
    "                self.nsemi += 1\n",
    "            else:\n",
    "                self.nneither += 1\n",
    "    \n",
    "    def nnodes(self):\n",
    "        ''' Return # nodes '''\n",
    "        return len(self.nodes)\n",
    "    \n",
    "    def nedges(self):\n",
    "        ''' Return # edges '''\n",
    "        return len(self.G)\n",
    "    \n",
    "    def hasEulerianWalk(self):\n",
    "        ''' Return true iff graph has Eulerian walk. '''\n",
    "        return self.nneither == 0 and self.nsemi == 2\n",
    "    \n",
    "    def hasEulerianCycle(self):\n",
    "        ''' Return true iff graph has Eulerian cycle. '''\n",
    "        return self.nneither == 0 and self.nsemi == 0\n",
    "    \n",
    "    def isEulerian(self):\n",
    "        ''' Return true iff graph has Eulerian walk or cycle '''\n",
    "        # technically, if it has an Eulerian walk\n",
    "        return self.hasEulerianWalk() or self.hasEulerianCycle()\n",
    "    \n",
    "    def eulerianWalkOrCycle(self):\n",
    "        ''' Find and return sequence of nodes (represented by\n",
    "            their k-1-mer labels) corresponding to Eulerian walk\n",
    "            or cycle '''\n",
    "        assert self.isEulerian()\n",
    "        g = self.G\n",
    "        if self.hasEulerianWalk():\n",
    "            g = g.copy()\n",
    "            g.setdefault(self.tail, []).append(self.head)\n",
    "        # graph g has an Eulerian cycle\n",
    "        tour = []\n",
    "        src = next(iter(g.keys())) # pick arbitrary starting node\n",
    "        \n",
    "        def __visit(n):\n",
    "            while len(g[n]) > 0:\n",
    "                dst = g[n].pop()\n",
    "                __visit(dst)\n",
    "            tour.append(n)\n",
    "        \n",
    "        __visit(src)\n",
    "        tour = tour[::-1][:-1] # reverse and then take all but last node\n",
    "            \n",
    "        if self.hasEulerianWalk():\n",
    "            # Adjust node list so that it starts at head and ends at tail\n",
    "            sti = tour.index(self.head)\n",
    "            tour = tour[sti:] + tour[:sti]\n",
    "        \n",
    "        # Return node list\n",
    "        return list(map(str, tour))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b5483ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DeBruijnGraph2(DeBruijnGraph):\n",
    "    def to_dot(self, weights=False):\n",
    "        ''' Return string with graphviz representation.  If 'weights'\n",
    "            is true, label edges corresponding to distinct k-1-mers\n",
    "            with weights, instead of drawing separate edges for\n",
    "            k-1-mer copies. '''\n",
    "        g = graphviz.Digraph(comment='DeBruijn graph')\n",
    "        for node in iter(self.G.keys()):\n",
    "            g.node(node.km1mer, node.km1mer)\n",
    "        for src, dsts in iter(self.G.items()):\n",
    "            if weights:\n",
    "                weightmap = {}\n",
    "                for dst in dsts:\n",
    "                    weightmap[dst] = weightmap.get(dst, 0) + 1\n",
    "                for dst, v in weightmap.items():\n",
    "                    g.edge(src.km1mer, dst.km1mer, label=str(v))\n",
    "            else:\n",
    "                for dst in dsts:\n",
    "                    g.edge(src.km1mer, dst.km1mer)\n",
    "        return g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0b25315",
   "metadata": {},
   "outputs": [],
   "source": [
    "def neighbors1mm(kmer, alpha):\n",
    "    ''' Generate all neighbors at Hamming distance 1 from kmer '''\n",
    "    neighbors = []\n",
    "    for j in range(len(kmer)-1, -1, -1):\n",
    "        oldc = kmer[j]\n",
    "        for c in alpha:\n",
    "            if c == oldc: continue\n",
    "            neighbors.append(kmer[:j] + c + kmer[j+1:])\n",
    "    return neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2b3dcede",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmerHist(reads, k):\n",
    "    ''' Return k-mer histogram and average # k-mer occurrences '''\n",
    "    kmerhist = {}\n",
    "    for read in reads:\n",
    "        for kmer in [ read[i:i+k] for i in range(len(read)-(k-1)) ]:\n",
    "            kmerhist[kmer] = kmerhist.get(kmer, 0) + 1\n",
    "    return kmerhist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1327b678",
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct1mm(read, k, kmerhist, alpha, thresh):\n",
    "    ''' Return an error-corrected version of read.  k = k-mer length.\n",
    "        kmerhist is kmer count map.  alpha is alphabet.  thresh is\n",
    "        count threshold above which k-mer is considered correct. '''\n",
    "    # Iterate over k-mers in read\n",
    "    for i in range(len(read)-(k-1)):\n",
    "        kmer = read[i:i+k]\n",
    "        # If k-mer is infrequent...\n",
    "        if kmerhist.get(kmer, 0) <= thresh:\n",
    "            # Look for a frequent neighbor\n",
    "            for newkmer in neighbors1mm(kmer, alpha):\n",
    "                if kmerhist.get(newkmer, 0) > thresh:\n",
    "                    # Found a frequent neighbor; replace old kmer\n",
    "                    # with neighbor\n",
    "                    read = read[:i] + newkmer + read[i+k:]\n",
    "                    break\n",
    "    # Return possibly-corrected read\n",
    "    return read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "765c5ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Non-greedy SCS, for comparison:\n",
    "def overlap(a, b, min_length=3):\n",
    "    ''' Return length of longest suffix of 'a' matching\n",
    "        a prefix of 'b' that is at least 'min_length'\n",
    "        characters long.  If no such overlap exists,\n",
    "        return 0. '''\n",
    "    start = 0  # start all the way at the left\n",
    "    while True:\n",
    "        start = a.find(b[:min_length], start)  # look for b's suffx in a\n",
    "        if start == -1:  # no more occurrences to right\n",
    "            return 0\n",
    "        # found occurrence; check for full suffix/prefix match\n",
    "        if b.startswith(a[start:]):\n",
    "            return len(a)-start\n",
    "        start += 1  # move just past previous match\n",
    "\n",
    "def scs(ss):\n",
    "    ''' Returns shortest common superstring of given\n",
    "        strings, which must be the same length '''\n",
    "    shortest_sup = None\n",
    "    for ssperm in itertools.permutations(ss):\n",
    "        sup = ssperm[0]  # superstring starts as first string\n",
    "        for i in range(len(ss)-1):\n",
    "            # overlap adjacent strings A and B in the permutation\n",
    "            olen = overlap(ssperm[i], ssperm[i+1], min_length=1)\n",
    "            # add non-overlapping portion of B to superstring\n",
    "            sup += ssperm[i+1][olen:]\n",
    "        if shortest_sup is None or len(sup) < len(shortest_sup):\n",
    "            shortest_sup = sup  # found shorter superstring\n",
    "    return shortest_sup  # return shortest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4fa78166",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pick_maximal_overlap(reads, k):\n",
    "    ''' Return a pair of reads from the list with a\n",
    "        maximal suffix/prefix overlap >= k.  Returns\n",
    "        overlap length 0 if there are no such overlaps.'''\n",
    "    reada, readb = None, None\n",
    "    best_olen = 0\n",
    "    for a, b in itertools.permutations(reads, 2):\n",
    "        olen = overlap(a, b, min_length=k)\n",
    "        if olen > best_olen:\n",
    "            reada, readb = a, b\n",
    "            best_olen = olen\n",
    "    return reada, readb, best_olen\n",
    "\n",
    "def greedy_scs(reads, k):\n",
    "    ''' Greedy shortest-common-superstring merge.\n",
    "        Repeat until no edges (overlaps of length >= k)\n",
    "        remain. '''\n",
    "    read_a, read_b, olen = pick_maximal_overlap(reads, k)\n",
    "    while olen > 0:\n",
    "        reads.remove(read_a)\n",
    "        reads.remove(read_b)\n",
    "        reads.append(read_a + read_b[-(len(read_b) - olen):])\n",
    "        read_a, read_b, olen = pick_maximal_overlap(reads, k)\n",
    "    return ''.join(reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "35cafb30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read sequences from FASTA file\n",
    "def read_fasta(file_path):\n",
    "    sequences = []\n",
    "    for record in SeqIO.parse(file_path, 'fasta'):\n",
    "        sequences.append(str(record.seq))\n",
    "\n",
    "    return sequences\n",
    "\n",
    "def visualize_overlap_graph(sequences, output_file, k):\n",
    "    '''Visualize overlap graph with Graphviz.'''\n",
    "    graph = graphviz.Digraph()\n",
    "    for seq in sequences:\n",
    "        graph.node(seq)\n",
    "\n",
    "    for seq1 in sequences:\n",
    "        for seq2 in sequences:\n",
    "            if seq1 != seq2:\n",
    "                overlap_len = overlap(seq1, seq2, k)\n",
    "                if overlap_len > 0:\n",
    "                    graph.edge(seq1, seq2, label=str(overlap_len))\n",
    "\n",
    "    graph.render(output_file, format='png', cleanup=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1c28c4a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AAGCTTGCGTTACCTAAAGG', 'AATGGATAGTGAGGTTAAGG', 'ACTAGATCCGAATGGATAGT', 'ACTAGCAAGCTTGCGTTACC', 'AGATCCGAATGGATAGTGAG', 'AGCAAGCTTGCGTTACCTAA', 'AGCACTAGCAAGCTTGCGTT', 'AGCTGAATGCGAGAAAGCAC', 'ATGCGAGAAAGCACTAGCAA', 'CCCCCCGAGAGCTGAATGCG', 'CCCGAGAGCTGAATGCGAGA', 'CGAGAAAGCACTAGCAAGCT', 'CTTGCGTTACCTAAAGGGCA', 'GAAAGCACTAGCAAGCTTGC', 'GAATCGAATGGATAGTGAGG', 'GAGAGCTGAATGCGAGAAAG', 'GCGTTACCTAAAGGGCAGGT', 'GGATAGTGAGGTTAAGGCCC', 'GGCCCCCCCGAGAGCTGAAT', 'GGTTAAGGCCCCCCCGAGAG', 'TAAGGCCCCCCCGAGAGCTG', 'TAGTGAGGTTAAGGCCCCCC', 'TCCGAATGGATAGTGAGGTT', 'TCGAATGGATAGTGAGGTTA', 'TGAATGCGAGAAAGCACTAG', 'TGAGGTTAAGGCCCCCCCGA', 'TTACCTAAAGGGCAGGTGGT']\n",
      "Reconstructed Sequence: ACTAGATCCGAATGGATAGTGAGGTTGAATCGAATGGATAGTGAGGTTAAGGCCCCCCCGAGAGCTGAATGCGAGAAAGCACTAGCAAGCTTGCGTTACCTAAAGGGCAGGTGGT\n"
     ]
    }
   ],
   "source": [
    "sequences = read_fasta(\"test.fa\")\n",
    "alpha = \"ACTG\"\n",
    "k = 7\n",
    "thresh = 3\n",
    "\n",
    "# Build de Bruijn graph\n",
    "kmer_histogram = kmerHist(sequences, k)\n",
    "corrected_sequences = [correct1mm(seq, k, kmer_histogram, alpha, thresh) for seq in sequences]\n",
    "dbg = DeBruijnGraph2(corrected_sequences, k)\n",
    "\n",
    "# Plot overlap graph\n",
    "dbg.to_dot().render(\"debruijn_graph\", format=\"png\", cleanup=True)\n",
    "visualize_overlap_graph(corrected_sequences, \"overlap_graph\", k)\n",
    "\n",
    "print(corrected_sequences)\n",
    "print(\"Reconstructed Sequence:\", greedy_scs(corrected_sequences.copy(), k))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
